subsys/net/lib/config/init.c

```diff
@@ -553,14 +580,539 @@ int net_config_init_app(const struct device *dev, const char *app_info)
 	return ret;
 }
 
+static struct net_if *get_interface(const struct net_init_config *config,
+				    int config_ifindex,
+				    const struct device *dev,
+				    const char **iface_name)
+{
+	const struct net_init_config_iface *cfg;
+	struct net_if *iface = NULL;
+	const char *name;
+
+	NET_ASSERT(IN_RANGE(config_ifindex, 0, ARRAY_SIZE(config->iface) - 1));
+
+	cfg = &config->iface[config_ifindex];
+
+	name = cfg->new_name;
+	if (name != NULL) {
+		iface = net_if_get_by_index(net_if_get_by_name(name));
+	}
+
+	if (iface == NULL) {
+		name = cfg->name;
+
+		if (name != NULL) {
+			iface = net_if_get_by_index(net_if_get_by_name(name));
+		}
+	}
+
+	if (iface == NULL) {
+		/* Get the interface by device */
+		const struct device *iface_dev;
+
+		name = cfg->dev;
+
+		iface_dev = device_get_binding(name);
+		if (iface_dev) {
+			iface = net_if_lookup_by_dev(iface_dev);
+		}
+	}
+
+	if (iface_name != NULL) {
+		*iface_name = name;
+	}
+
+	return iface;
+}
+
+static void ipv6_setup_from_generated(struct net_if *iface,
```

Comment:
Do we really need separate code paths for setting up devices from YAML vs. from Kconfig? Why not creating instances of the configuration structs when the legacy Kconfig approach is in use? This consumes a little more RAM but the maintenance cost would be a lot less due to reduced code complexity and less bugs.

Plus I believe that "_from_generated" indicates a violation of the information hiding principle. I don't think the code should know anything about how the configs came about even if it's only in the naming.

Reply:
Kconfig will only support a subset of what can be represented in Yaml and only one instance. But why should the Yaml not be backwards compatible with Kconfig? IMO this is a must-have - see my above point about decoupling of configuration format from configuration content plus we need a clear migration path.

If that's the case then everything in the code should be represented in a configuration format agnostic way and treated in the same way. In the Kconfig case you can let cmake include a header that uses the exact same C header format you're generating from Yaml - just inserting the data via Kconfig constants. In other words, this would be a very cheap Kconfig-to-C-header generator backend.

Once you have this everything else should work just the same for any backend approach. Otherwise this approach doesn't scale IMO.

Reply:
I did not convert existing Kconfig handling code to use the same C struct format as the yaml handling code, as this felt quite pointless because the we would just use extra flash without apparent benefits.

Reply:
It might be a trade off, I agree. The question is whether it is really so clear that the few extra bytes we need for config structs will not be easily offset by the simplification of the code which IMO is an apparent benefit because it takes maintenance burden and error potential off the community, makes the code easier to extend plus it may well be that more efficient code will make up for the bytes lost to the declaration of the config. Did you actually compare the two approaches in terms of RAM/ROM usage? To me this seems very close to premature optimization as compared to the complexity burden and legacy we'll have to carry along for a long time.

Reply:
And still the naming is a clear indicator of a broken encapsulation. It just doesn't make sense from an architectural pov. It's clearly a code smell.